metafeed-finder.js

```diff
@@ -0,0 +1,229 @@
+const pull = require('pull-stream')
+const debug = require('debug')('ssb:replication-scheduler')
+const detectSsbNetworkErrorSeverity = require('ssb-network-errors')
+const { where, type, live, toPullStream } = require('ssb-db2/operators')
+const { validateMetafeedAnnounce } = require('ssb-meta-feeds/validate')
+
+const DEFAULT_PERIOD = 500
+
+module.exports = class MetafeedFinder {
+  constructor(ssb, opts, period) {
+    this._ssb = ssb
+    this._opts = opts
+    this._period = period || DEFAULT_PERIOD
+    this._map = new Map() // mainFeedId => rootMetaFeedId
+    this._inverseMap = new Map() // rootMetaFeedId => mainFeedId
+    this._requestsByMainfeedId = new Map() // mainFeedId => Array<Calback>
+    this._latestRequestTime = 0
+    this._timer = null
+
+    // If at least one hops template is configured, then load
+    if (
+      this._opts.partialReplication &&
+      Object.values(this._opts.partialReplication).some((templ) => !!templ)
+    ) {
+      if (!this._ssb.db || !this._ssb.db.query) {
+        throw new Error(
+          'ssb-replication-scheduler expects ssb-db2 to be installed, to use partial replication'
+        )
+      }
+      this._loadAllFromLog()
+    }
+  }
+
+  fetch(mainFeedId, cb) {
+    if (this._map.has(mainFeedId)) {
+      const metaFeedId = this._map.get(mainFeedId)
+      cb(null, metaFeedId)
+      return
+    } else if (mainFeedId === this._ssb.id) {
+      this._ssb.metafeeds.find((err, rootMF) => {
+        if (err) {
+          return cb(err)
+        } else if (!rootMF) {
+          cb(null, null)
+        } else {
+          const metaFeedId = rootMF.keys.id
+          this._map.set(mainFeedId, metaFeedId)
+          this._inverseMap.set(metaFeedId, mainFeedId)
+          cb(null, metaFeedId)
+        }
+      })
+    } else {
+      this._request(mainFeedId, cb)
+    }
+  }
+
+  get(mainFeedId) {
+    return this._map.get(mainFeedId)
+  }
+
+  getInverse(metaFeedId) {
+    return this._inverseMap.get(metaFeedId)
+  }
+
+  liveStream() {
+    return pull(
+      this._ssb.db.query(
+        where(type('metafeed/announce')),
+        live(),
+        toPullStream()
+      ),
+      pull.filter(this._validateMetafeedAnnounce),
+      pull.map((msg) => {
+        const [mainFeedId, metaFeedId] = this._pluckFromAnnounceMsg(msg.value)
+        this._map.set(mainFeedId, metaFeedId)
+        this._inverseMap.set(metaFeedId, mainFeedId)
+        return [mainFeedId, metaFeedId]
+      })
+    )
+  }
+
+  _loadAllFromLog() {
+    pull(
+      this._ssb.db.query(where(type('metafeed/announce')), toPullStream()),
+      pull.filter(this._validateMetafeedAnnounce),
+      pull.drain(
+        (msg) => {
+          const [mainFeedId, metaFeedId] = this._pluckFromAnnounceMsg(msg.value)
+          this._map.set(mainFeedId, metaFeedId)
+          this._inverseMap.set(metaFeedId, mainFeedId)
+        },
+        () => {
+          debug(
```

Comment:
When we are done here, I think it would be safer if this module set up the live stream mapping and then req-manager could use a simpler version with no map. Right now this module depends on req-manager behaving good in order to have a consistent cache.

Reply:
Can you clarify what do you mean with "req-manager with no map"? And the consistent cache thing?

Reply:
I was thinking metafeed-finder would start up a `liveStream` here so that there is no "time" between this end of "old" and starting the live stream that will update the cache. That is what I mean with consistent.

Reply:
Ah I see, yeah there has to be a tighter time "gap" between old and live drainage. Thanks for clarifying